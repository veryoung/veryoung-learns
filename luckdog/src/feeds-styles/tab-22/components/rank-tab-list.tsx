/* eslint-disable react-native/no-unused-styles */
import React, { useState } from 'react';
import { View, Text, StyleSheet } from '@tencent/hippy-react-qb';
import FeedsTheme from '../../../framework/FeedsTheme';
import { FeedsUIStyle } from '../../../framework/FeedsConst';

type Tab = {
  sGroupName: string;
  sGroupId: string;
};

export enum RankTabType {
  /** 全宽 tab */
  BIG = 'BIG',
  /** 小 tab，放在标题位置 */
  SMALL = 'SMALL',
}

type Props = {
  /** tab 列表 */
  tabList: Tab[];
  /** tab 样式类型 */
  tabType: RankTabType;
  /** 选中 tab 的索引 */
  initSelectedIndex: number;
  /** tab 点击事件函数 */
  onTabClick?: (index: number) => void;
  /** layout 事件函数 */
  onLayout?: (event: Record<string, any>) => void;
};

const getTabStyle = (index: number, length: number, styles: Record<string, any>) => {
  if (index === 0) return styles.tabWithLeftBorderRadius;

  if (index === length - 1) {
    return {
      ...styles.tabSpace,
      ...styles.tabWithRightBorderRadius,
    };
  }

  return styles.tabSpace;
};

export const RankTabList = ({ tabType, tabList, initSelectedIndex = 0, onTabClick, onLayout }: Props) => {
  if (!tabList?.length) return null;

  const [selectedIndex, setSelectedIndex] = useState(initSelectedIndex);

  const styles = stylesMap[tabType];

  const onClick = (index: number): void => {
    setSelectedIndex(index);
    onTabClick?.(index);
  };

  return (
    <View style={styles.wrap} onLayout={onLayout}>
      {tabList.map(({ sGroupName, sGroupId }, index) => (
        <View
          key={sGroupId}
          style={[
            styles.tab,
            getTabStyle(index, tabList.length, styles),
            index === selectedIndex ? styles.tabSelected : null,
          ]}
          onClick={() => onClick(index)}
        >
          <Text style={[styles.tabName, index === selectedIndex ? styles.tabNameSelected : null]}>{sGroupName}</Text>
        </View>
      ))}
    </View>
  );
};

const smallTabStyles = StyleSheet.create({
  wrap: {
    flex: 1,
    flexDirection: 'row',
  },
  tab: {
    paddingLeft: 10,
    paddingRight: 10,
    paddingTop: 5,
    paddingBottom: 5,
    colors: FeedsTheme.SkinColor.N1,
    backgroundColors: FeedsTheme.SkinColor.N8,
    height: 28,
    justifyContent: 'center',
    alignItems: 'center',
  },
  tabName: {
    fontSize: 14,
    colors: FeedsTheme.SkinColor.N1,
    opacity: 0.7,
  },
  tabSpace: {
    marginLeft: 1,
  },
  tabSelected: {
    backgroundColors: FeedsTheme.SkinColor.N2,
  },
  tabNameSelected: {
    opacity: 1,
  },
  tabWithLeftBorderRadius: {
    borderTopLeftRadius: 6,
    borderBottomLeftRadius: 6,
  },
  tabWithRightBorderRadius: {
    borderTopRightRadius: 6,
    borderBottomRightRadius: 6,
  },
});

const bigTabStyles = StyleSheet.create({
  wrap: {
    flexDirection: 'row',
    marginHorizontal: FeedsUIStyle.T1,
    paddingBottom: 20,
  },
  tab: {
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColors: FeedsTheme.SkinColor.N8,
    height: 36,
  },
  tabName: {
    fontSize: FeedsUIStyle.T1_5,
    colors: FeedsTheme.SkinColor.N1,
    opacity: 0.7,
  },
  tabSpace: {
    marginLeft: 2,
  },
  tabSelected: {
    backgroundColors: FeedsTheme.SkinColor.N2,
  },
  tabNameSelected: {
    opacity: 1,
  },
  tabWithLeftBorderRadius: {
    borderTopLeftRadius: 18,
    borderBottomLeftRadius: 18,
  },
  tabWithRightBorderRadius: {
    borderTopRightRadius: 18,
    borderBottomRightRadius: 18,
  },
});

const stylesMap = {
  [RankTabType.BIG]: bigTabStyles,
  [RankTabType.SMALL]: smallTabStyles,
};
