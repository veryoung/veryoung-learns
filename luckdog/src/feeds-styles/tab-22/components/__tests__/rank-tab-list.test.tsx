import React from 'react';
import renderer from 'react-test-renderer';

import { RankTabList, RankTabType } from '../rank-tab-list';
import { commonProps } from '@test/mocks/props';
import { TAB_LIST } from './mocks/rank-tab-list';

describe('<RankTabList />', () => {
  describe('render testing', () => {
    it('render normal', () => {
      const props = {
        ...commonProps,
        tabList: TAB_LIST,
        tabType: RankTabType.SMALL,
        initSelectedIndex: 0,
      };
      const tree = renderer.create(<RankTabList {...props} />);
      expect(tree).toMatchSnapshot();
    });
  });
});
