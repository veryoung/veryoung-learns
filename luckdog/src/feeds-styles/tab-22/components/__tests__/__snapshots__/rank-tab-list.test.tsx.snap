// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RankTabList /> render testing render normal 1`] = `
<div
  nativeName="View"
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
    }
  }
>
  <div
    nativeName="View"
    onClick={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColors": Array [
            "#F5F7F9",
            "#202327",
            "#F5F7F9",
            "#202327",
          ],
          "colors": Array [
            "#000000",
            "#686D74",
            "#000000",
            "#ffffff",
          ],
          "height": 28,
          "justifyContent": "center",
          "paddingBottom": 5,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 5,
        },
        Object {
          "borderBottomLeftRadius": 6,
          "borderTopLeftRadius": 6,
        },
        Object {
          "backgroundColors": Array [
            "#EFBF8A",
            "#B0906C",
            "#EFBF8A",
            "#B0906C",
          ],
        },
      ]
    }
  >
    <p
      nativeName="Text"
      style={
        Array [
          Object {
            "colors": Array [
              "#000000",
              "#686D74",
              "#000000",
              "#ffffff",
            ],
            "fontSize": 14,
            "opacity": 0.7,
          },
          Object {
            "opacity": 1,
          },
        ]
      }
      text="推荐榜"
    >
      推荐榜
    </p>
  </div>
  <div
    nativeName="View"
    onClick={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColors": Array [
            "#F5F7F9",
            "#202327",
            "#F5F7F9",
            "#202327",
          ],
          "colors": Array [
            "#000000",
            "#686D74",
            "#000000",
            "#ffffff",
          ],
          "height": 28,
          "justifyContent": "center",
          "paddingBottom": 5,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 5,
        },
        Object {
          "marginLeft": 1,
        },
        null,
      ]
    }
  >
    <p
      nativeName="Text"
      style={
        Array [
          Object {
            "colors": Array [
              "#000000",
              "#686D74",
              "#000000",
              "#ffffff",
            ],
            "fontSize": 14,
            "opacity": 0.7,
          },
          null,
        ]
      }
      text="热度榜"
    >
      热度榜
    </p>
  </div>
  <div
    nativeName="View"
    onClick={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColors": Array [
            "#F5F7F9",
            "#202327",
            "#F5F7F9",
            "#202327",
          ],
          "colors": Array [
            "#000000",
            "#686D74",
            "#000000",
            "#ffffff",
          ],
          "height": 28,
          "justifyContent": "center",
          "paddingBottom": 5,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 5,
        },
        Object {
          "marginLeft": 1,
        },
        null,
      ]
    }
  >
    <p
      nativeName="Text"
      style={
        Array [
          Object {
            "colors": Array [
              "#000000",
              "#686D74",
              "#000000",
              "#ffffff",
            ],
            "fontSize": 14,
            "opacity": 0.7,
          },
          null,
        ]
      }
      text="飙升榜"
    >
      飙升榜
    </p>
  </div>
  <div
    nativeName="View"
    onClick={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColors": Array [
            "#F5F7F9",
            "#202327",
            "#F5F7F9",
            "#202327",
          ],
          "colors": Array [
            "#000000",
            "#686D74",
            "#000000",
            "#ffffff",
          ],
          "height": 28,
          "justifyContent": "center",
          "paddingBottom": 5,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 5,
        },
        Object {
          "borderBottomRightRadius": 6,
          "borderTopRightRadius": 6,
          "marginLeft": 1,
        },
        null,
      ]
    }
  >
    <p
      nativeName="Text"
      style={
        Array [
          Object {
            "colors": Array [
              "#000000",
              "#686D74",
              "#000000",
              "#ffffff",
            ],
            "fontSize": 14,
            "opacity": 0.7,
          },
          null,
        ]
      }
      text="完本榜"
    >
      完本榜
    </p>
  </div>
</div>
`;
